repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,437 @@
+              validToken.defaultRegion.map(dr => OpenStackServiceHeader.DEFAULT_REGION.toString -> dr).foreach(addtlReqHdrs.+=)
+              validToken.contactId.map(ci => OpenStackServiceHeader.CONTACT_ID.toString -> ci).foreach(addtlReqHdrs.+=)
+              validToken.impersonatorId.map(iid => OpenStackServiceHeader.IMPERSONATOR_ID.toString -> iid).foreach(addtlReqHdrs.+=)
+              validToken.impersonatorName.map(in => OpenStackServiceHeader.IMPERSONATOR_NAME.toString -> in).foreach(addtlReqHdrs.+=)
+
+              // If configured, add roles header
+              if (config.getIdentityService.isSetRolesInHeader) {
+                addtlReqHdrs += (OpenStackServiceHeader.ROLES.toString -> validToken.roles.mkString(","))
+              }
+
+              // Extract the tenant and add the x-authorization header
+              val uriTenantOption = requestHandler.extractTenant(request.getRequestURI)
+              uriTenantOption match {
+                case Some(uriTenant) =>
+                  addtlReqHdrs += (OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString -> s"$X_AUTH_PROXY $uriTenant")
+                case None =>
+                  addtlReqHdrs += (OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString -> X_AUTH_PROXY)
+              }
+
+              // Construct and add the tenant id header
+              val tenantsToPass = buildTenantHeader(config.getTenantHandling, validToken.defaultTenantId, validToken.tenantIds, uriTenantOption)
+              addtlReqHdrs += (OpenStackServiceHeader.TENANT_ID.toString -> tenantsToPass.mkString(","))
+
+              // If configured, authorize the URI tenant against the token
+              val authorizedTenant = requestHandler.tenantAuthorized(uriTenantOption, validToken) match {
+                case Some(Failure(e)) =>
+                  Reject(SC_UNAUTHORIZED, failure = Some(e))
+                case _ =>
+                  Pass
+              }
+
+              // Authorize token endpoints and, if configured, populate the x-catalog header
+              val endpoints = authorizedTenant match {
+                case Pass =>
+                  requestHandler.handleEndpoints(authToken, validToken) match {
+                    case Some(Success(endpointsData)) =>
+                      //If I'm configured to put the endpoints into a x-catalog do it
+                      if (config.getIdentityService.isSetCatalogInHeader) {
+                        // todo: don't check this flag twice
+                        addtlReqHdrs += (PowerApiHeader.X_CATALOG.toString -> Base64.encodeBase64String(endpointsData.endpointsJson.getBytes))
+                        Pass
+                      } else {
+                        Pass
+                      }
+                    case Some(Failure(x)) =>
+                      //Reject them with 403
+                      Reject(SC_FORBIDDEN, failure = Some(x))
```

Comment:
This is real bad. In the case where a failure is caused by a bad admin token, this should return a 500 not a 403.